### PR TITLE
Light client p2p protocol

### PR DESCRIPTION
**Pull request by lontivero**
This is for keeping some progress and discussions about the coming light client (bip 157 and bip 158). Note that Bitcoin Core has a first on-review PR for Golomb-Rice filters (https://github.com/bitcoin/bitcoin/pull/12254).

The idea is starting with the low-hanging fruit such as the more relevant p2p protocol message. Once it is compatible with btcd nodes, we could use it in WasabiWallet (far future)

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Wow this is awesome! will read more on this.

**Comment by lontivero** [comment]
Bitcoin Core team members are discussing about what info to include into the filters, how many filter types should exist and how to verify the filters are correct. Additionally depending on the amount and type of data, the P value (used to set the FP rate) could need to be recalculated.
https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-May/015972.html

**Comment by knocte** [comment]
Just curious, did BIP157 and BIP158 implementations ever got included into BitcoinCore?

**Comment by lontivero** [comment]
> Just curious, did BIP157 and BIP158 implementations ever got included into BitcoinCore?

* BIP158: **yes**
* BIP157: **no**

File: NBitcoin/Protocol/Payloads/GetCompactFiltersPayload.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NBitcoin.Protocol
+{
+	public enum FilterType : byte
+	{
+		Basic = (0x00),
+		Extended = (0x01)
```
Comment (dangershony) [inline NBitcoin/Protocol/Payloads/GetCompactFiltersPayload.cs]: What does extended filter hold?

Comment (lontivero) [reply NBitcoin/Protocol/Payloads/GetCompactFiltersPayload.cs]: https://github.com/bitcoin/bips/commit/b65ca1c094c1efb193a79d39efac6e34d390254a#diff-367f1ff45d7efb5c6774e4112cd17489

**Comment by lontivero** [comment]
@NicolasDorier I would like to continue with this PR but I hit a wall that kept me blocked for months. I have implemented the p2p messages and I can see the network traffic is okay and that Bitcoin Core receives and replay okay so, that part is correct and works perfect.

The problem I have is I am not sure how to implement the compact filters and compact filter headers. Considering the bip 157 recommends to synchronize the block headers first, should implement this using something like the `CompactFiltersChainBehavior`? or should I modify the `ChainBehavior`? Do you think it is correct to reuse the `SlimChain` to keep the chain of compact filter headers?

I would appreciate if you could take a look at the bip 157 and provide me some guidance. 

----------

Btw, the commits that starts with `[WIP]` are just garbage but I pushed them to give you an idea of what I am talking about.

**Comment by lontivero** [comment]
It seems the fact this PR is 401 has bad karma and will not be authorized. cuack! 

Idk when i will find the time to continue with this so, what do you think about removing the [WIP] commits? Would you merge it?

**Comment by lontivero** [comment]
@NicolasDorier I hate to break your zen but as every year, since 2018, I feel the need to go back to this PR. 

**What do you think about removing the [WIP] commits? Would you merge it?** Thank you.

**Comment by NicolasDorier** [comment]
well I completely forgot this PR, if you think it's ready to merge let's merg eit.

File: NBitcoin.Tests/ProtocolTests.cs [hunk 2]
```diff
@@ -342,6 +342,80 @@ public void CanHandshakeWithSeveralTemplateBehaviors()
 				}
 			}
 		}
+
+		//[Fact]
+		//[Trait("Protocol", "Protocol")]
```
Comment (NicolasDorier) [inline NBitcoin.Tests/ProtocolTests.cs:347]: Remove this so the CI can test it.

**Comment by lontivero** [comment]
The PR contains only the protocol messages (payload). I removed the `CompactFiltersChainBehavior` and the integration test because I don't know how to implement it correctly.

**Comment by knocte** [comment]
@lontivero just curious, how far is this for a light-node to not depend on, for exampe, electrum servers?

**Comment by lontivero** [comment]
It is still far from that because the bip recommends to download the block headers chain first and then download the compact filters chain.

You can take a look at this two commits in order to have an idea about how to make a naive implementation of this:
* https://github.com/lontivero/NBitcoin/commit/aee05b855212d225f435e6b9bfa24fdddad71355
* https://github.com/lontivero/NBitcoin/commit/a1f348517f90b6b6cbb03ff7df3babb5a8488e05

**Pull request merged**